### PR TITLE
Quic fix

### PR DESCRIPTION
**Pull request by mr-torgue**
### 1. Why is this pull request needed and what does it do?
This pull request fixes two things for DNS over QUIC (DoQ):
1. Enabled early data for 0-RTT
2. Fixed DoQ issue where a DoQ response UDP size is not set to 65535

Functions have been tested with unit tests and in-field testing.

**Early Data:**
To enable early data, quic-go has to use `ListenEarly` instead of `Listen`. This has been changed in `server_quic.go`. Note, this does bring some risks: early data is vulnerable against replay attacks and does not provide non-repudiation. However, it does increase performance which is important in the case of DNS.

**Size Issues**
EDNS defines an OPT record that can be used to advertise a UDP size. In the case of DoQ, this size should be ignored and a response should set the size to 65535, as indicated in [RFC9250](https://www.rfc-editor.org/rfc/rfc9250.html#name-message-sizes) CoreDNS does not do this. This is caused by the `Size` function in `plugin/pkg/edns/edns.go` and the `Proto` function. It only recognises udp and tcp. If tcp is used the size is correctly set to 65535, however QUIC uses udp, so the EDNS udpsize in the query is also used in the response.

I considered two ways of solving this.
1. Rewriting the incoming query
2. Adding a Proto function to ResponseWriter instances that returns the protocol (such as quic)

Option 2 seemed the cleanest, however since the ResponseWriter interface does not enforce a Proto function it is implemented ad hoc.
If possible it would be cleaner to add a `Proto` function to this interface so that every Writer has to implement such a function. 

### 2. Which issues (if any) are related?
I have not created issues on GitHub. See description above for the issues.

### 3. Which documentation changes (if any) need to be made?
None.

### 4. Does this introduce a backward incompatible change or deprecation?
None. However, 0RTT early data does introduce some risk.

### CODE REVIEW

File: request/request.go [hunk 1]
```diff
@@ -112,6 +112,10 @@ func (r *Request) LocalAddr() string { return r.W.LocalAddr().String() }
 
 // Proto gets the protocol used as the transport. This will be udp or tcp.
 func (r *Request) Proto() string {
+	// return Write.Proto(), if it is implemented
+	if protoProvider, ok := r.W.(interface{ Proto() string }); ok {
+		return protoProvider.Proto()
```
Comment (houyuwushang) [inline request/request.go:117]: This value also becomes the default outbound transport in `plugin/pkg/proxy.lookupDNS`: when neither `ForceTCP` nor `PreferUDP` is set, it assigns `state.Proto()` to `dns.Client.Net`. A normal DoQ request through `forward` therefore passes `"quic"` to miekg/dns, and every default plain-DNS lookup fails with `dial quic: unknown network quic`.

I reproduced this with a real UDP `dnstest` upstream: current `master` passed 20/20, while this head and the current GitHub merge commit failed 20/20. Please separate the inbound transport label from the supported outbound network selection (or explicitly map DoQ there) and add a DoQ-to-forward regression test.

File: plugin/plugin.go [hunk 2]
```diff
@@ -134,6 +134,21 @@ func (pw *pluginWriter) TsigTimersOnly(b bool) { pw.ResponseWriter.TsigTimersOnl
 // Hijack implements dns.ResponseWriter.
 func (pw *pluginWriter) Hijack() { pw.ResponseWriter.Hijack() }
 
+// Proto gets the protocol used as the transport. This will be udp or tcp.
+func (pw *pluginWriter) Proto() string {
+	// return Write.Proto(), if it is implemented
+	if protoProvider, ok := pw.ResponseWriter.(interface{ Proto() string }); ok {
```
Comment (houyuwushang) [inline plugin/plugin.go:140]: The optional method is not preserved by existing response-writer wrappers. For example, `header.ResponseHeaderWriter` embeds `dns.ResponseWriter`, whose interface has no `Proto()`, so this assertion falls back to its `UDPAddr`. In a DoQ -> header -> downstream probe on the current merge commit, downstream observed `"udp"` in 20/20 runs.

This makes protocol-dependent behavior depend on plugin ordering and can also hide or reveal the `forward` failure above. Please use a transport signal that survives writer wrapping, or make propagation enforced across in-tree wrappers, and add a real wrapper-chain regression.

**Review by houyuwushang — CHANGES_REQUESTED** [review]
I verified both intended behaviors independently on the current merge commit: a real resumed QUIC connection used 0-RTT, and an EDNS=512 DoQ response used a 65535-byte truncation budget. Both also passed under `-race`. The two inline findings are integration blockers introduced by exposing `"quic"` through `Request.Proto()`.

Please add regressions for DoQ -> forward and a real in-tree writer-wrapper chain. The current tests only prove that the optional `Proto()` method works when every test wrapper explicitly forwards it, which misses both failures.

For 0-RTT, please add an actual session-resumption test and exercise the early-data opcode boundary in [RFC 9250 section 4.5](https://www.rfc-editor.org/rfc/rfc9250.html#section-4.5). This is a replay/privacy-sensitive default behavior change and would be easier to assess separately from the message-size fix. Also, [section 4.6](https://www.rfc-editor.org/rfc/rfc9250.html#section-4.6) says the EDNS UDP payload size is ignored; it does not require advertising 65535 in the response, so the PR description should match the truncation behavior implemented here. If `"quic"` remains a public `Proto()` result, tcp/udp-only comments and docs need updating.

The visible lint (13 findings) and DCO checks also still need fixing.

File: plugin/plugin.go (hunk 2 above)
Comment (mr-torgue) [reply plugin/plugin.go:140]: Agree, this would work if Proto is enforced by the ResponseWriter (which it is not).
I will open an issue for this because I have not found a clean way to do this yet.
